mart/callbacks/anomalib.py

```diff
@@ -0,0 +1,425 @@
+    **kwargs,
+):
+    image_adv = image.clone()
+
+    # Return image to [0, 1] space
+    if image_mean is not None and image_std is not None:
+        image_adv = image_adv * image_std[..., None, None] + image_mean[..., None, None]
+
+    # Rotate image by angle degrees
+    if angle is not None:
+        if mode == "three_pass":
+            theta = torch.deg2rad(angle)
+            image_adv = rotate_three_pass(
+                image_adv, theta, N=-1, padding_mode=padding_mode
+            )
+        else:
+            image_adv = rotate(image_adv, angle, mode=mode, padding_mode=padding_mode)
+
+    # Convert image from RGB to HSV space
+    if hue is not None or sat is not None:
+        image_hsv = rgb_to_hsv(image_adv)
+        image_hue, image_sat, image_val = torch.unbind(image_hsv, dim=-3)
+
+    # Additive hue perturbation with STE clipping
+    if hue is not None:
+        image_hue = image_hue + hue[:, None, None]
+        image_hue = torch.remainder(image_hue, 2 * torch.pi)
+
+    # Additive saturation perturbation with STE clipping
+    if sat is not None:
+        image_sat = image_sat + sat[:, None, None]
+        image_sat = image_sat + (torch.clip(image_sat, 0.0, 1.0) - image_sat).detach()
+
+    # Convert image fro HSV to RGB space
+    if hue is not None or sat is not None:
+        image_hsv = torch.stack([image_hue, image_sat, image_val], dim=-3)
+        image_adv = hsv_to_rgb(image_hsv)
+
+    # Re-normalize image to image_mean and image_std
+    if image_mean is not None and image_std is not None:
+        image_adv = (image_adv - image_mean[..., None, None]) / image_std[
+            ..., None, None
+        ]
+
+    return {"benign_image": image, "image": image_adv}
+
+
+# FIXME: Make this a @staticmethod
+@torch.no_grad()
+def perturb_mask(
+    mask,
+    *,
+    angle,
+    mode="nearest",  # bilinear | nearest | three_pass
+    padding_mode="zeros",  # constant | replicate | reflect | circular | zeros | border | reflection
+    **kwargs,
+):
+    mask_rot = mask.clone()
+    mask_rot = mask_rot[..., None, :, :].float()
+
+    if mode == "three_pass":
+        theta = torch.deg2rad(angle)
+        mask_rot = rotate_three_pass(mask_rot, theta, N=-1, padding_mode=padding_mode)
+    else:
+        mask_rot = rotate(mask_rot, angle, mode=mode, padding_mode=padding_mode)
+
+    mask_rot = mask_rot[..., 0, :, :].int()
+
+    return {
+        "benign_mask": mask,
+        "mask": mask_rot,
+    }
+
+
+# FIXME: Make this a @staticmethod
+def compute_loss(anomaly_maps, mask, **kwargs):
+    true_negatives = mask == 0
+    negative_loss = -torch.sum(anomaly_maps * true_negatives, dim=(1, 2)) / (
+        torch.sum(true_negatives, dim=(1, 2)) + 1e-8
+    )
+
+    true_positives = mask == 1
+    positive_loss = torch.sum(anomaly_maps * true_positives, dim=(1, 2)) / (
+        torch.sum(true_positives, dim=(1, 2)) + 1e-8
+    )
+
+    loss = negative_loss + positive_loss
+
+    # decrease negatives and increase positives
```

Comment:
Given that you removed using `maximize=True` in Adam, it may be worth adding an extra comment here that this will be minimized, and the intended effect is exactly opposite than that of the comment.

Reply:
I assume losses are made to be minimized and gains are made to be maximized, but fixed in 85e85bb37907b1a2c3432ee073117786d0a7206a, I think.